python/semantic_kernel/contents/history_reducer/chat_history_summarization_reducer.py

```diff
@@ -0,0 +1,226 @@
+# Copyright (c) Microsoft. All rights reserved.
+
+import logging
+import sys
+from typing import Any
+
+from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
+from semantic_kernel.utils.experimental_decorator import experimental_class
+
+if sys.version < "3.11":
+    from typing_extensions import Self  # pragma: no cover
+else:
+    from typing import Self  # type: ignore # pragma: no cover
+
+from pydantic import Field
+
+from semantic_kernel.connectors.ai.chat_completion_client_base import ChatCompletionClientBase
+from semantic_kernel.const import DEFAULT_SERVICE_NAME
+from semantic_kernel.contents.chat_history import ChatHistory
+from semantic_kernel.contents.chat_message_content import ChatMessageContent
+from semantic_kernel.contents.history_reducer.chat_history_reducer import ChatHistoryReducer
+from semantic_kernel.contents.history_reducer.chat_history_reducer_utils import (
+    SUMMARY_METADATA_KEY,
+    contains_function_call_or_result,
+    extract_range,
+    locate_safe_reduction_index,
+    locate_summarization_boundary,
+)
+from semantic_kernel.exceptions.content_exceptions import ChatHistoryReducerException
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_SUMMARIZATION_PROMPT = """
+Provide a concise and complete summarization of the entire dialog that does not exceed 5 sentences.
+
+This summary must always:
+- Consider both user and assistant interactions
+- Maintain continuity for the purpose of further dialog
+- Include details from any existing summary
+- Focus on the most significant aspects of the dialog
+
+This summary must never:
+- Critique, correct, interpret, presume, or assume
+- Identify faults, mistakes, misunderstanding, or correctness
+- Analyze what has not occurred
+- Exclude details from any existing summary
+"""
+
+
+@experimental_class
+class ChatHistorySummarizationReducer(ChatHistoryReducer):
+    """A ChatHistory with logic to summarize older messages past a target count."""
+
+    service: ChatCompletionClientBase
+    summarization_instructions: str = Field(
+        default_factory=lambda: DEFAULT_SUMMARIZATION_PROMPT,
+        description="The summarization instructions.",
+    )
+    use_single_summary: bool = Field(True, description="Whether to use a single summary message.")
+    fail_on_error: bool = Field(True, description="Raise error if summarization fails.")
+    service_id: str = Field(
```

Comment:
can't we get service_id from the service? or vice versa?

Reply:
What if you have multiple execution settings defined for the same ChatCompletion class? service_id is optional, so if one wants to provide it, we'll use it, otherwise fall back to "default."